### PR TITLE
Fixes #1632: http2 decoder fix for crash generated by AFL fuzzer.

### LINKED ISSUE #1632: http2 decoder crashes on input from AFL fuzzer
```
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44                                                                                                                                                                                                                                                                                                               
44	      return INTERNAL_SYSCALL_ERROR_P (ret) ? INTERNAL_SYSCALL_ERRNO (ret) : 0;
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007fbe934a86d3 in __pthread_kill_internal (threadid=<optimized out>, signo=6) at pthread_kill.c:78
#2  0x00007fbe9344fc4e in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007fbe93437902 in __GI_abort () at abort.c:79
#4  0x00007fbe9343781e in __assert_fail_base (fmt=0x7fbe935c3b98 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x4c54bf "length > 0", file=file@entry=0x4b7b48 "/home/gmurthy/opensource/skupper-router/src/decoders/http2/http2_decoder.c", line=line@entry=249, function=function@entry=0x4cc030 <__PRETTY_FUNCTION__.3.lto_priv.10> "decompress_headers") at assert.c:94
#5  0x00007fbe93447d87 in __assert_fail (assertion=0x4c54bf "length > 0", file=0x4b7b48 "/home/gmurthy/opensource/skupper-router/src/decoders/http2/http2_decoder.c", line=249, function=0x4cc030 <__PRETTY_FUNCTION__.3.lto_priv.10> "decompress_headers") at assert.c:103
#6  0x0000000000445751 in decompress_headers (decoder=0x217e1098, stream_id=0, data=0x217ec5c5 "", length=0) at /home/gmurthy/opensource/skupper-router/src/decoders/http2/http2_decoder.c:249
#7  0x0000000000445bb0 in get_request_headers (decoder=0x217e1098) at /home/gmurthy/opensource/skupper-router/src/decoders/http2/http2_decoder.c:342
#8  0x0000000000445e48 in parse_request_header (decoder=0x217e1098, data=0x7ffcb804c9c8, length=0x7ffcb804c9c0) at /home/gmurthy/opensource/skupper-router/src/decoders/http2/http2_decoder.c:387
#9  0x000000000044692b in decode (conn_state=0x217e1090, from_client=true, data=0x217eba10 "", length=15) at /home/gmurthy/opensource/skupper-router/src/decoders/http2/http2_decoder.c:584
#10 0x00000000004252fd in LLVMFuzzerTestOneInput (data=0x217eb9b0 "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n", size=111) at /home/gmurthy/opensource/skupper-router/tests/fuzz/fuzz_http2_decoder.c:55
#11 0x00000000004b3d5a in main (argc=58, argv=0x217d8d70) at /home/gmurthy/opensource/skupper-router/tests/fuzz/StandaloneFuzzTargetMain.c:126
(gdb) 

```

### PR DESCRIPTION
**Pull request by ganeshmurthy**

### CODE REVIEW

**Comment by ted-ross** [comment]
I think this is a problem:

There should be no asserts in any observer code.  No correctness can be assumed.  All of these conditions must be handled as errors which stop the observer from parsing.

```
$ grep assert -r src/decoders/
src/decoders/http2/http2_decoder.c:    assert(scratch_buffer->size != 0);
src/decoders/http2/http2_decoder.c:    assert(data_length > 0);
src/decoders/http2/http2_decoder.c:    assert(conn_state);
src/decoders/http2/http2_decoder.c:    assert(length > 0);
src/decoders/http2/http2_decoder.c:    assert(bit < 8);
src/decoders/http2/http2_decoder.c:    assert(scratch_buffer->size > 0);
src/decoders/http1/http1_decoder.c:#include <assert.h>
src/decoders/http1/http1_decoder.c:    assert(config);
src/decoders/http1/http1_decoder.c:    assert(hconn);
src/decoders/http1/http1_decoder.c:    assert(decoder->buffer.length > 0);
src/decoders/http1/http1_decoder.c:    assert(hrs);
src/decoders/http1/http1_decoder.c:        assert(DEQ_HEAD(hconn->hrs_queue) == hrs);  // expect completed in-order
src/decoders/http1/http1_decoder.c:    assert(!decoder->hrs);   // state machine violation
src/decoders/http1/http1_decoder.c:    assert(hrs);
src/decoders/http1/http1_decoder.c:    assert(decoder->hdr_content_length);
src/decoders/http1/http1_decoder.c:        assert(decoder->hrs);
src/decoders/http1/http1_decoder.c:    assert(decoder->hrs);
```

File: src/decoders/http2/http2_decoder.c [hunk 1]
```diff
@@ -338,6 +346,28 @@ bool get_request_headers(qd_http2_decoder_t *decoder)
...
+    //
+    // Before the call to decompress_headers(), we need to make sure that there is some data left in the scratch buffer before we decompress it.
+    //
+    int buffer_data_size = scratch_buffer->size - scratch_buffer->offset;
+    int contains_pad_length = scratch_buffer->size - pad_length;
+    int pad_length_offset = scratch_buffer->size - pad_length - scratch_buffer->offset;
+    bool valid_pad_length = contains_pad_length > 0;
+    bool valid_buffer_data = buffer_data_size > 0;
+    bool valid_pad_length_offset = pad_length_offset > 0;
+    if(decoder->frame_payload_length == 0 || !valid_buffer_data || !valid_pad_length || !valid_pad_length_offset) {
```
Comment (ted-ross) [inline src/decoders/http2/http2_decoder.c:358]: 'if' is not a function call.  Please put spaces after ifs (both styles are found throughout this file).

Comment (ganeshmurthy) [reply src/decoders/http2/http2_decoder.c]: Replaced `if(` with `if (`

**Comment by ganeshmurthy** [comment]
> There should be no asserts in any observer code

I have removed the relevant asserts. Please take a look

File: src/decoders/http2/http2_decoder.c [hunk 2]
```diff
@@ -143,7 +149,7 @@ qd_http2_frame_type get_frame_payload_length_and_type(qd_http2_decoder_t *decode
 
 static int allocate_move_to_scratch_buffer(qd_decoder_buffer_t  *scratch_buffer, const uint8_t *data, size_t capacity)
 {
-    if (capacity > HTTP2_SCRATCH_BUFFER_MAX_SIZE) {
+    if (capacity > HTTP2_SCRATCH_BUFFER_MAX_SIZE || capacity <= 0) {
```
Comment (ted-ross) [inline src/decoders/http2/http2_decoder.c:152]: capacity is an unsigned quantity.  It can't be < 0.

File: src/decoders/http2/http2_decoder.c [hunk 3]
```diff
@@ -158,7 +164,9 @@ static int allocate_move_to_scratch_buffer(qd_decoder_buffer_t  *scratch_buffer,
 
 static int reallocate_move_to_scratch_buffer(qd_decoder_buffer_t  *scratch_buffer, const uint8_t *data, size_t data_length)
 {
-    assert(scratch_buffer->size != 0);
+    if (scratch_buffer->size == 0 || data_length <= 0) {
```
Comment (ted-ross) [inline src/decoders/http2/http2_decoder.c:167]: Ditto on the < 0 for unsigned data_length.

File: src/decoders/http2/http2_decoder.c [hunk 4]
```diff
@@ -246,7 +251,10 @@ void qd_http2_decoder_connection_free(qd_http2_decoder_connection_t *conn_state)
  */
 static int decompress_headers(qd_http2_decoder_t *decoder, uint32_t stream_id, const uint8_t *data, size_t length)
 {
-    assert(length > 0);
+    if (length <= 0) {
+        qd_log(LOG_HTTP2_DECODER, QD_LOG_ERROR, "[C%"PRIu64"] decompress_headers - passed in length < 0", decoder->conn_state->conn_id);
```
Comment (ted-ross) [inline src/decoders/http2/http2_decoder.c:255]: I don't think this should be logged at the ERROR level, maybe DEBUG or not at all.  We don't want an external process to be able to clog up our logs by sending garbage protocol traffic.

File: src/decoders/http2/http2_decoder.c [hunk 5]
```diff
@@ -246,7 +251,10 @@ void qd_http2_decoder_connection_free(qd_http2_decoder_connection_t *conn_state)
  */
 static int decompress_headers(qd_http2_decoder_t *decoder, uint32_t stream_id, const uint8_t *data, size_t length)
 {
-    assert(length > 0);
+    if (length <= 0) {
```
Comment (ted-ross) [inline src/decoders/http2/http2_decoder.c:254]: length is an unsigned quantity and cannot be less than 0.

**Review by ted-ross — CHANGES_REQUESTED** [review]

**Comment by ganeshmurthy** [comment]
@ted-ross I think I have fixed everything. Please take a look.

**Review by ted-ross — APPROVED** [review]

**Pull request merged**